Enable single-pass BBQ baseline JIT by default

ddegazio opened the pull request:
#### e3f0b033dcf4ed2f7b03c7c359acf4ab11a4c08c
<pre>
Enable single-pass BBQ baseline JIT by default
<a href="https://bugs.webkit.org/show_bug.cgi?id=252209">https://bugs.webkit.org/show_bug.cgi?id=252209</a>
rdar://100332177

Reviewed by Mark Lam and Yusuke Suzuki.

The new BBQ baseline JIT is largely complete, passes all of our stress tests,
and doesn&apos;t introduce a performance regression. This patch moves to enable it
as our default implementation for the BBQ tier, supplanting the Air backend.

One notable hole in the new JIT&apos;s support is in some of the recent WASM extensions:
typed function references, tail calls, and GC. Since these are still incomplete
currently, and disabled by default, single-pass BBQ doesn&apos;t yet support them.
When these features are manually enabled, single-pass BBQ JIT will be disabled
currently, and defer back to Air. Air and B3 can also be selected by passing
`--useSinglePassBBQJIT=0` in the command line options.

* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: <a href="https://commits.webkit.org/261153@main">https://commits.webkit.org/261153@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/527e89571aa7d85be30b6f5208077f5fffcf9270

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110751 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19833 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/43305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2096 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/119575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114717 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21240 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10944 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/1780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116497 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/43305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/103106 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/43305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44184 "layout-tests (failure)") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/99405 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/12440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/43305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10538 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/13006 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8969 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100480 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18384 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/43305 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31346 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7728 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/14961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108509 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26752 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me

MenloDorian commented on Source/JavaScriptCore/runtime/Options.cpp:
```diff
@@ -665,6 +665,18 @@ void Options::notifyOptionsChanged()
+            // FIXME: Add WASM typed function references support to single-pass BBQ JIT. https://bugs.webkit.org/show_bug.cgi?id=253191
+            // FIXME: Add WASM tail calls support to single-pass BBQ JIT. https://bugs.webkit.org/show_bug.cgi?id=253192
+            Options::useSinglePassBBQJIT() = false;
+        }
```
Can you add a check after this that disables Air/B3 BBQ if `useSinglePassBBQJIT` is true?

ddegazio replied:
`useSinglePassBBQJIT` supersedes both of the other backends on its own, so this would be unnecessary.

MenloDorian commented on Tools/Scripts/run-jsc-stress-tests:
```diff
@@ -1915,14 +1921,15 @@ def runWebAssemblySpecTestBase(mode, specHarnessPath, postfix, *optionalTestSpec
+        runWithOutputHandler("wasm-air" + (postfix || ''), noisyOutputHandler, specHarnessJsPath, "--useWasmLLInt=false", "--useRandomizingExecutableIslandAllocation=true", "--useSinglePassBBQJIT=false", *(FTL_OPTIONS + optionalTestSpecificOptions))
+        if $isFTLPlatform
+            runWithOutputHandler("wasm-bbq" + (postfix || ''), noisyOutputHandler, specHarnessJsPath, "--useWasmLLInt=false", *(FTL_OPTIONS + optionalTestSpecificOptions))
+            runWithOutputHandler("wasm-b3" + (postfix || ''), noisyOutputHandler, specHarnessJsPath, "--useWasmLLInt=false", "--wasmBBQUsesAir=false", "--useSinglePassBBQJIT=false", *(FTL_OPTIONS + optionalTestSpecificOptions))
```
wasm-bbq should not be dependent on FTL.

ddegazio replied:
BBQ itself is not dependent on B3/FTL, at least not moreso than Air (we use a few tidbits like `B3::ValueRep` but this shouldn't be a limiting dependency). The reason it's under `$isFTLPlatform` is mostly to be conservative with where we enable it - so far, it's been tested on 64-bit ARM and Intel, which corresponds to the platforms covered by `$isFTLPlatform`. We can look into enabling it on more platforms in the future.

ddegazio replied:
Added an explanatory comment in the latest push (well, actually to the first occurrence, in `runWebAssembly`).

MenloDorian reviewed (approved):
r=me too

ddegazio committed:
Enable single-pass BBQ baseline JIT by default
https://bugs.webkit.org/show_bug.cgi?id=252209
rdar://100332177

Reviewed by Mark Lam and Yusuke Suzuki.

The new BBQ baseline JIT is largely complete, passes all of our stress tests,
and doesn't introduce a performance regression. This patch moves to enable it
as our default implementation for the BBQ tier, supplanting the Air backend.

One notable hole in the new JIT's support is in some of the recent WASM extensions:
typed function references, tail calls, and GC. Since these are still incomplete
currently, and disabled by default, single-pass BBQ doesn't yet support them.
When these features are manually enabled, single-pass BBQ JIT will be disabled
currently, and defer back to Air. Air and B3 can also be selected by passing
`--useSinglePassBBQJIT=0` in the command line options.

* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Tools/Scripts/run-jsc-stress-tests:

Canonical link: https://commits.webkit.org/261153@main
